File: programmers/Courses/알고리즘-문제-해설/5.py
#
# Created by habi on 2023-08-01.
#
# Problem
# https://school.programmers.co.kr/learn/courses/18/lessons/1880
#


def solution(land):
    answer = 0

    # mat의 index가 len(land)가 넘어가도 문제가 없도록 +1 해줌
    mat = [[0 for i in range(4)] for j in range(len(land) + 1)]

    for j in range(len(land)):
        # land에 접근할 때는 길이에서 -1 해줘야함
        idx = (len(land) - 1) - j

        # mat의 아래에서부터 각 지점까지의 최댓값을 계산하기 위해 역으로 처리함
        # mat의 가장 아래는 어차피 0으로 채워져있어 값 계산에 영향을 주지 않음
        for k in range(4):
            mat[idx][k] = (
                max(
                    mat[idx + 1][(k + 1) % 4],
                    mat[idx + 1][(k + 2) % 4],
                    mat[idx + 1][(k + 3) % 4],
                )
                + land[idx][k]
            )

    answer = max(mat[0])

    return answer


def solution2(land):
    answer = 0

    mat = [0 for i in range(4)]

    for row in land:
        temp = mat.copy()

        for i in range(4):
            mat[i] = max(temp[:i] + temp[i + 1 :]) + row[i]

    answer = max(mat)

    return answer

File: programmers/Courses/알고리즘-문제-해설/5.cpp
//
// Created by habi on 2023-08-02.
//
// Problem
// https://school.programmers.co.kr/learn/courses/18/lessons/1880
//

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int solution(vector<vector<int>> land) {
  int answer = 0;

  vector<int> vec(4);
  vector<vector<int>>::iterator iter;

  for (iter = land.begin(); iter != land.end(); iter++) {
    vector<int> row = *iter;
    vector<int> temp(4);
    copy(vec.begin(), vec.end(), temp.begin());

    for (int i = 0; i < 4; i++) {
      // vector temp로부터 해당 index를 제외한 나머지 elements만 남긴다.
      int save = temp[i];
      temp.erase(temp.begin() + i);

      // 최댓값과 row[index] 값의 합을 vector vec의 index 위치에 저장한다.
      vec[i] = *max_element(temp.begin(), temp.end()) + row[i];

      // vector temp에 원래 값 insert
      temp.insert(temp.begin() + i, save);
    }
  }

  answer = *max_element(vec.begin(), vec.end());

  return answer;
}
